packages/astro/src/virtual-modules/i18n.ts

```diff
@@ -191,3 +209,184 @@ export const getPathByLocale = (locale: string) => I18nInternals.getPathByLocale
+ *
+ * ```js
+ * import { pathHasLocale } from "astro:i18n";
+ * getLocaleByPath("italiano"); // returns `true`
+ * getLocaleByPath("es"); // returns `true`
+ * getLocaleByPath("it-VT"); // returns `false`
+ * ```
+ */
+export const pathHasLocale = (path: string) => I18nInternals.pathHasLocale(path, locales);
+
+/**
+ *
+ * This function returns a redirect to the default locale configured in the
+ *
+ * @param {APIContext} context The context passed to the middleware
+ * @param {ValidRedirectStatus?} statusCode An optional status code for the redirect.
+ */
+export let redirectToDefaultLocale: (
+	context: APIContext,
+	statusCode?: ValidRedirectStatus
+) => Response | undefined;
+
+if (i18n?.routing === 'manual') {
+	redirectToDefaultLocale = I18nInternals.redirectToDefaultLocale({
+		base,
+		trailingSlash,
+		format,
+		defaultLocale,
+		locales,
+		routing: strategy,
+		domains,
+		fallback,
+	});
+} else {
+	redirectToDefaultLocale = noop('redirectToDefaultLocale');
+}
+/**
+ *
+ * Use this function to return a 404 when:
+ * - the current path isn't a root. e.g. / or /<base>
+ * - the URL doesn't contain a locale
+ *
+ * When a `Response` is passed, the new `Response` emitted by this function will contain the same headers of the original response.
+ *
+ * @param {APIContext} context The context passed to the middleware
+ * @param {Response?} response An optional `Response` in case you're handling a `Response` coming from the `next` function.
+ *
+ */
+export let notFound: (context: APIContext, response?: Response) => Response | undefined;
+
+if (i18n?.routing === 'manual') {
+	notFound = I18nInternals.notFound({
+		base,
+		trailingSlash,
+		format,
+		defaultLocale,
+		locales,
+		routing: strategy,
+		domains,
+		fallback,
+	});
+} else {
+	notFound = noop('notFound');
+}
+
+/**
+ * Checks whether the current URL contains a configured locale. Internally, this function will use `APIContext#url.pathname`
+ *
+ * @param {APIContext} context The context passed to the middleware
+ */
+export let requestHasLocale: (context: APIContext) => boolean;
+
+if (i18n?.routing === 'manual') {
+	requestHasLocale = I18nInternals.requestHasLocale(locales);
+} else {
+	requestHasLocale = noop('requestHasLocale');
+}
+
+/**
+ * Allows to use the build-in fallback system of Astro
+ *
+ * @param {APIContext} context The context passed to the middleware
+ * @param {Response?} response An optional `Response` in case you're handling a `Response` coming from the `next` function.
+ */
+export let useFallback: UseFallback;
+
+if (i18n?.routing === 'manual') {
+	useFallback = I18nInternals.useFallback({
+		base,
+		trailingSlash,
+		format,
+		defaultLocale,
+		locales,
+		routing: strategy,
+		domains,
+		fallback,
+	});
+} else {
+	useFallback = noop('useFallback');
+}
+
+// @ematipico: This is a helper type because, because I don't know how to extract from TS
+type Options = {
+	prefixDefaultLocale?: boolean;
+	redirectToDefaultLocale?: boolean;
+};
```

Comment:
I'm not 100% sure what you're trying to extract here, can't you just do `AstroConfig['i18n'][option]`?

Reply:
No, because the type is more complex than that, and it's above my skills. 

`AstroConfig['i18n']['routing']` can be undefined, and I can manage by using `NonNullable`, although `routing` is now an object OR a literal `"manual"`. I just want to extract the object.